cluster-autoscaler/cloudprovider/kwok/kwok_provider_test.go

```diff
@@ -156,6 +157,110 @@ func TestNodeGroups(t *testing.T) {
 	})
 }
 
+func TestRefresh(t *testing.T) {
+	fakeClient := &fake.Clientset{}
+	var nodesFrom string
+	fakeClient.Fake.AddReactor("get", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+		getAction := action.(core.GetAction)
+
+		if getAction == nil {
+			return false, nil, nil
+		}
+
+		if getAction.GetName() == defaultConfigName {
+			if nodesFrom == "configmap" {
+				return true, &apiv1.ConfigMap{
+					Data: map[string]string{
+						configKey: testConfig,
+					},
+				}, nil
+			}
+
+			return true, &apiv1.ConfigMap{
+				Data: map[string]string{
+					configKey: testConfigDynamicTemplates,
+				},
+			}, nil
+
+		}
+
+		if getAction.GetName() == defaultTemplatesConfigName {
+			if nodesFrom == "configmap" {
+				return true, &apiv1.ConfigMap{
+					Data: map[string]string{
+						templatesKey: testTemplates,
+					},
+				}, nil
+			}
+		}
+
+		return true, nil, errors.NewNotFound(apiv1.Resource("configmaps"), "whatever")
+	})
+
+	os.Setenv("POD_NAMESPACE", "kube-system")
+
+	t.Run("refresh nodegroup target size", func(t *testing.T) {
+		nodesFrom = "configmap"
+		ngName := "kind-worker"
+		fakeNodeLister := newTestAllNodeLister(map[string]*apiv1.Node{
+			"node1": {
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+					Labels: map[string]string{
+						"kwok-nodegroup": ngName,
+					},
+				},
+			},
+			"node2": {
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node2",
+					Labels: map[string]string{
+						"kwok-nodegroup": ngName,
+					},
+				},
+			},
+			"node3": {
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node3",
+					Labels: map[string]string{
+						"kwok-nodegroup": ngName,
+					},
+				},
+			},
+			"node4": {
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node4",
+				},
+			},
+		})
+
+		ko := &kwokOptions{
+			kubeClient:      fakeClient,
+			autoscalingOpts: &config.AutoscalingOptions{},
+			discoveryOpts:   &cloudprovider.NodeGroupDiscoveryOptions{},
+			resourceLimiter: cloudprovider.NewResourceLimiter(
+				map[string]int64{cloudprovider.ResourceNameCores: 1, cloudprovider.ResourceNameMemory: 10000000},
+				map[string]int64{cloudprovider.ResourceNameCores: 10, cloudprovider.ResourceNameMemory: 100000000}),
+			allNodesLister: fakeNodeLister,
+			ngNodeListerFn: testNodeLister,
+		}
+
+		p, err := BuildKwokProvider(ko)
+		assert.NoError(t, err)
+		assert.NotNil(t, p)
+
+		err = p.Refresh()
```

Comment:
`kwokConfig.status` is coming out nil here for some reason because of which `node4` test case is not throwing an error

Reply:
<img width="1920" alt="image" src="https://github.com/kubernetes/autoscaler/assets/5654791/6f998df5-4a34-406f-a354-31a2e53153a2">

Do you mean the `p.config.status` is nil? I ran this test case in debug mode and found that this field was not nil. Is this the expected value?

Reply:
Sorry, it was a misunderstanding on my end. I ran the test again. Looks good to me.